Make LinearPackedParams works with both torchscript and torch.package

ShijunK opened the pull request:
Summary:
Customized `__getstate__`/`__setstate__` didn't call super (torch.nn.Module), and won't restore attributes (e.g. `_modules`) after being serialized and deserialized via torch.package

The patch will use saved `__dict__` if available, or restore init states for attributes

Test Plan: `buck test caffe2/test/:quantization -- test_linear_api`

Differential Revision: D33711086

ShijunK commented on torch/nn/modules/module.py:
```diff
@@ -1183,7 +1183,7 @@ def __getattr__(self, name: str) -> Union[Tensor, 'Module']:
                 return modules[name]
         raise AttributeError("'{}' object has no attribute '{}'".format(
-            type(self).__name__, name))
+            type(self), name))
```
removed, it was modified for debugging, to distinguish two types with same name but from different modules.

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -83,13 +84,43 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        if '_modules' not in self.__dict__:
+            self._modules = OrderedDict()
 
-    @torch.jit.export
```
why are we moving this? is this not required now?

ShijunK replied:
Is it required? I see sample at https://github.com/pytorch/pytorch/blob/79a258f4484f97fab954e699e5d2b5f2dae6548e/torch/csrc/jit/docs/serialization.md#__getstate__-and-__setstate__   which does not use export annotation. 

And, there is `checkScriptable` as part of `test_linear_api`, which doesn't break after removing the export annotation.

ShijunK replied:
I tried to put export annotation back, now the jit script related test fails
```
> RuntimeError: 
Module 'LinearPackedParams' has no attribute '__dict__' :
  File "/data/users/shijunk/fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantization#binary,link-tree/torch/nn/quantized/modules/linear.py", line 91
        qweight, bias = self._weight_bias()
        # Make a copy and save attributes from torch.nn.Module
        state = self.__dict__.copy()
                ~~~~~~~~~~~~~ <--- HERE
        return qweight, bias, self.training, self.dtype, state
```

@gmagogsfm , is above error expected? that in torchscript, we cannot use `self.__dict__`?

jerryzh168 replied:
I think this is probably expected..
but I'm also surprised that we can remove this decorator, I don't exactly remember why we have this in the beginning, @gmagogsfm could you comment on whether this is needed?

ShijunK replied:
cc @jamesr66a, I see originally intorduced `@torch.jit.export` as part of #24048, which itself is a work around issue #24045 , can you comment on if `@torch.jit.export` is still required for `__{get|set}state__`?

gmagogsfm replied:
> I tried to put export annotation back, now the jit script related test fails
> 
> ```
> > RuntimeError: 
> Module 'LinearPackedParams' has no attribute '__dict__' :
>   File "/data/users/shijunk/fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantization#binary,link-tree/torch/nn/quantized/modules/linear.py", line 91
>         qweight, bias = self._weight_bias()
>         # Make a copy and save attributes from torch.nn.Module
>         state = self.__dict__.copy()
>                 ~~~~~~~~~~~~~ <--- HERE
>         return qweight, bias, self.training, self.dtype, state
> ```
> 
> @gmagogsfm , is above error expected? that in torchscript, we cannot use `self.__dict__`?

Yeah, this is expected. `__dict__` isn't implemented in TS

gmagogsfm replied:
> I think this is probably expected.. but I'm also surprised that we can remove this decorator, I don't exactly remember why we have this in the beginning, @gmagogsfm could you comment on whether this is needed?

I think this decorator is needed. Otherwise, get/set_state won't be compiled/serialized by TS. Could we verify if everything still works (including save/load) without decorator? If so, then that would be quite interesting

ShijunK replied:
> > I think this is probably expected.. but I'm also surprised that we can remove this decorator, I don't exactly remember why we have this in the beginning, @gmagogsfm could you comment on whether this is needed?
> 
> I think this decorator is needed. Otherwise, get/set_state won't be compiled/serialized by TS. Could we verify if everything still works (including save/load) without decorator? If so, then that would be quite interesting

In test/quantization/core/test_quantized_module.py, line 111 there is one step to verify torch script (including scripting, tracing, save/load) 
```
self.checkScriptable(qlinear_copy, [[X_q]], check_save_load=True)
```
The actual function is implemented in `torch/testing/_internal/common_quantization.py`

From my limited knowledge of torch.script, it looks like covering the scenario?

gmagogsfm replied:
Hi @ShijunK,

I did some testing, if `__setstate__` and `__getstate__` do not exist or not decorated by `export`, then they are not called when loading with `torch.jit.load`

I also checked the test you mentioned (line 111 of `test_quantized_module.py`), it is indeed checking that modules can be saved/loaded and can produce same results. 

So the only plausible possibility is that in those tests, the `state` that should have been restored do not make a difference in module's behavior. I do not fully understand this module's behavior, maybe you could verify if that's the case.

jerryzh168 replied:
I see, I think the reason the test did not fail is because we didn't set a random weight/bias for quantized linear: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L107, we are saving and loading the default quantized linear module

cc @ShijunK could you help add a `set_weight_bias` call after the initialization and set a random quantized weight?

ShijunK replied:
> I see, I think the reason the test did not fail is because we didn't set a random weight/bias for quantized linear: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L107, we are saving and loading the default quantized linear module
> 
> cc @ShijunK could you help add a `set_weight_bias` call after the initialization and set a random quantized weight?

add weight and bias setting before testing torch jit scriptable. 

also, after a few iteration, found that the underlying packed param torchbind class already support pickle (pack/unpack) wegith, bias. no need to hack nn module's `__{set|get}state__` anymore. It also remove the need to hack `copy/deepcopy` too. 

submitted a new PR ver to complete remove all these overrides.

jerryzh168 commented:
can you still add the changes to qlinaer_copy to set_weight_bias before calling checkScriptable: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L109

jerryzh168 reviewed (approved):
Looks great, thanks!

ShijunK commented:
> can you still add the changes to qlinaer_copy to set_weight_bias before calling checkScriptable: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L109

yes, the additional step is still here in the test. I didn't remove it. this will help to make sure we test qlinear with random weight/bias, not testing against default init values.

ShijunK committed:
Make LinearPackedParams works with both torchscript and torch.package (#71656)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71656

Customized `__getstate__`/`__setstate__` didn't call super (torch.nn.Module), and won't restore attributes (e.g. `_modules`) after being serialized and deserialized via torch.package

After a few iteration, as it turns out, pack/unpack linear param has been supported in torchbind class already, no need to hack torch module anymore.

Test Plan: `buck test caffe2/test/:quantization -- test_linear_api`

Reviewed By: jerryzh168

Differential Revision: D33711086

fbshipit-source-id: c0faceb4352e5272f5bebc66dea2fdd7b43d5eb4

ShijunK commented:
> 2022-02-07T16:08:46.0588780Z RuntimeError: test_spectral_ops failed!

most likely due to upstream change in https://github.com/librosa/librosa/issues/1114

ShijunK commented:
test_distribution failure show up in other PRs too:
https://github.com/pytorch/pytorch/pull/72431
https://github.com/pytorch/pytorch/pull/72430

not related to this change, ignore
